lib/links/parser.rb

```diff
@@ -0,0 +1,52 @@
+require 'httparty'
+
+class Parser
+
+    def initialize(url)
+        @elements = {a: ["href"],
+                img: ["src"],
+                script: ["src"],
+                frame: ["src"],
+                iframe: ["src"],
+                link: ["href"],
+                source: ["src", "srcset"],
+                track: ["src"] }
+
+        @url = url
+    end
+
+    def run
+        all_links = Array.new
+        @doc = Nokogiri::HTML(HTTParty.get(@url))
+
+        @elements.each do |elem, attr_list|
+            elems = get_elems(elem)
+
+            attr_list.each do |attr|
+                links = extract_links(elems, attr)
+                all_links = (all_links << links).flatten
+            end
+        end
+
+        return all_links.uniq
+    end
+
+    def get_elems(elem)
+        return @doc.css(elem)
+    end
+
+    def extract_links(elems, attr)
+        all_elem_links = elems.map{ |e| e[attr]}.compact
+        relatives, absolutes = all_elem_links.partition{ |link| is_relative?(link)}
+        joined_relatives = relatives.map{ |rel_link| URI.join(@url, rel_link).to_s}
+        return (joined_relatives << absolutes).flatten
```

Comment:
is this a replacement to the similar code you have on #1 ?

If so, it could be expressed a lot more clearly. might be worth looking into my comments there :)

To give some concrete examples:
* newlines and spacing are important to isolate ideas. Much like writing a text with different paragraphs
* chaining things together instead of using temporary variables at every step:

```ruby
# hard to read
def extract_links
  foo = array.map { ... }
  bar = foo.filter { ... }

  return bar
end

# easier to read, exactly the same functionality
def extract_links
  array.
    map { ... }.
    filter { ... }
end
```